feat: Add "Comet Fuzz" fuzz-testing utility

andygrove opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

N/A

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

Comet Fuzz is a standalone project for generating random data and queries and executing queries against Spark 
with Comet disabled and enabled and checking for incompatibilities.

Although it is a simple tool it has already been useful in finding many bugs, including:

- #479 
- #480 
- #481 
- #482 
- #483 
- #484 
- #485 
- #486 

Comet Fuzz is inspired by the [SparkFuzz](https://ir.cwi.nl/pub/30222) paper from Databricks and CWI.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

New standalone `comet-fuzz` project.

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->


Manually tested.

andygrove commented on dev/scalastyle-config.xml:
```diff
@@ -242,7 +242,7 @@ This file is divided into 3 sections:
       <parameter name="groups">java,scala,org,apache,3rdParty,comet</parameter>
       <parameter name="group.java">javax?\..*</parameter>
       <parameter name="group.scala">scala\..*</parameter>
-      <parameter name="group.org">org\.(?!apache\.comet).*</parameter>
```
We should exclude all apache packages from the `org` group, not just comet. We have a separate rule below for excluding comet from the apache group.

Without this change it was not possible to satisfy both scalastyle and scalafix at the same time as they wanted different import orders.

kazuyukitanimura commented on fuzz-testing/pom.xml:
```diff
@@ -0,0 +1,106 @@
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.scala-lang</groupId>
+            <artifactId>scala-library</artifactId>
+            <version>${scala.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.rogach</groupId>
+            <artifactId>scallop_${scala.binary.version}</artifactId>
+            <version>5.1.0</version>
+        </dependency>
```
nit: (optional) we use `<dependencyManagement>` for version

kazuyukitanimura commented on fuzz-testing/src/main/scala/org/apache/comet/fuzz/DataGen.scala:
```diff
@@ -0,0 +1,153 @@
+
+    // TODO add examples of all supported types, including complex types
+    val dataTypes = Seq(
+      (DataTypes.ByteType, 0.2),
```
What about adding boolean and binary types?

andygrove commented:
Thanks for the review @kazuyukitanimura. I have addressed your comments and also added support for decimal type. I  removed the TODO comments and added a roadmap section to the README to cover some of the planned future work.

andygrove commented:
I'm not sure why tests are being cancelled on this PR.

viirya commented:
> I'm not sure why tests are being cancelled on this PR.

I saw the same on some PRs too. Re-triggering CIs can make them back.

viirya commented on fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:
```diff
@@ -0,0 +1,121 @@
+    val table = spark.table(tableName)
+
+    val func = Utils.randomChoice(Meta.aggFunc, r)
+    val args = Range(0, func.num_args)
+      .map(_ => Utils.randomChoice(table.columns, r))
+
+    val groupingCols = Range(0, 2).map(_ => Utils.randomChoice(table.columns, r))
```
Do we need to exclude gouping columns from aggregation function arguments?

andygrove replied:
If it is valid SQL I think we should allow it

viirya commented on fuzz-testing/src/main/scala/org/apache/comet/fuzz/QueryGen.scala:
```diff
@@ -0,0 +1,121 @@
+      s"$rightColProjection " +
+      s"FROM $leftTableName l " +
+      s"$joinType JOIN $rightTableName r " +
+      s"ON l.$leftCol = r.$rightCol " +
```
`leftCol` and `rightCol` are randomly picked up. Are they always same data types?

viirya replied:
Or it doesn't matter and we let Spark automatically cast between them?

andygrove replied:
Exactly. We just pick random columns and let Spark add implicit casts when needed.

viirya commented on fuzz-testing/src/main/scala/org/apache/comet/fuzz/Main.scala:
```diff
@@ -0,0 +1,88 @@
+
+  lazy val spark: SparkSession = SparkSession
+    .builder()
+    .master("local[*]")
```
Do we need to hardcode this?

andygrove replied:
I have removed this. We specify master during spark-submit so do not need to hardcode it here.

andygrove commented:
Thanks for the review @viirya
